test/retained_topic_map.h

```diff
@@ -0,0 +1,199 @@
+//
+// Created by wkl04 on 12-12-2019.
+//
+
+#ifndef MQTTSUBSCRIPTION_RETAINED_TOPIC_MAP_H
+#define MQTTSUBSCRIPTION_RETAINED_TOPIC_MAP_H
+
+#include <mqtt/string_view.hpp>
+#include <map>
+#include "path_tokenizer.h"
+
+template<typename Value >
+class retained_topic_map
+{
+    using node_id_type = size_t;
+    using path_entry_key = std::pair< node_id_type, std::string>;
+
+    static constexpr node_id_type root_node_id = 0;
+
+    struct path_entry
+    {
+        node_id_type id;
+        uint32_t count = 1;
```

Comment:
Why use uint32_t specifically? Not size_t?

Reply:
Because size_t is architecture dependent and uint32_t is not. I guess when you want to make some statements about how many entries can be stored in the subscription map, you should use a defined count size. Now the maximum number of entries is 4 billion, otherwise it is architecture dependent.

Reply:
Ok.

But path_entry is already going to be sizeof(uint64_t)*2 because of struct padding.

Reply:
you could say we optimise for 64 bit and make it always 64 bits ...

Reply:
I guess it makes sense to go for 64 bits, most server machines are 64 bits anyway, and the count gets depleted much later. Do we need an exception on when count is depleted ? I guess that would be correct

Reply:
It's not about making it specifically 64bit hardware optimized.

It's about struct packing.

On a 32bit platform, the compiler is going to lay this struct out as 2-32bit integers. On a 64bit platform, the compiler is going to lay this struct out as 1-64bit, and 2-32bits. One of those 32bits will be accessible to your program, and the other will be hidden.

If you simply make both types as size_t, you get the optimal outcome.

There's no reason to concern ourselves with the differences in functionality being different depending on what hardware you're running on. That's always going to be true. But if we're already paying for that extra 32bits of space for the alignment / struct padding, why wouldn't we just use it for the count?